exonum/src/helpers/fabric/password.rs

```diff
@@ -0,0 +1,168 @@
+pub enum SecretKeyType {
+    Consensus,
+    Service,
+}
+
+impl PassInputMethod {
+    /// Get passphrase using selected method.
+    /// Details of this process differs for different secret key types and whether we run node
+    /// or generate config files.
+    pub fn get_passphrase(self, key_type: SecretKeyType, run: bool) -> ZeroizeOnDrop<String> {
+        match self {
+            PassInputMethod::Terminal => {
+                let prompt = match key_type {
+                    SecretKeyType::Consensus => "Enter consensus key passphrase",
+                    SecretKeyType::Service => "Enter service key passphrase",
+                };
+                prompt_passphrase(prompt, run).expect("Failed to read password from stdin")
+            }
+            PassInputMethod::EnvVariable(name) => {
+                let var = if let Some(ref name) = name {
+                    name
+                } else {
+                    match key_type {
+                        SecretKeyType::Consensus => "EXONUM_CONSENSUS_PASS",
+                        SecretKeyType::Service => "EXONUM_SERVICE_PASS",
+                    }
+                };
+                ZeroizeOnDrop(env::var(var).unwrap_or_else(|e| {
+                    panic!("Failed to get password from env variable: {}, {}", var, e)
+                }))
+            }
+            PassInputMethod::CmdLineParameter(pass) => pass,
+        }
+    }
+}
+
+impl Default for PassInputMethod {
+    fn default() -> Self {
+        PassInputMethod::Terminal
+    }
+}
+
+impl FromStr for PassInputMethod {
+    type Err = failure::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "" {
+            return Ok(Default::default());
+        }
+
+        if s == "stdin" {
+            return Ok(PassInputMethod::Terminal);
+        }
+
+        if let Some(prefix) = s.get(0..3) {
+            if prefix == "env" {
+                let env_var = s.split(':').nth(1).map(|s| s.to_owned());
+                return Ok(PassInputMethod::EnvVariable(env_var));
+            }
+        }
+
+        if let Some(prefix) = s.get(0..4) {
+            if prefix == "pass" {
+                let pass = s.split(':').nth(1).unwrap_or("");
+                return Ok(PassInputMethod::CmdLineParameter(ZeroizeOnDrop(
+                    pass.to_owned(),
+                )));
+            }
+        }
+
+        bail!("Failed to parse passphrase input method");
+    }
+}
+
+fn prompt_passphrase(prompt: &str, run: bool) -> io::Result<ZeroizeOnDrop<String>> {
+    if run {
+        return Ok(ZeroizeOnDrop(read_password_from_tty(Some(prompt))?));
+    }
+
+    loop {
+        let password = ZeroizeOnDrop(read_password_from_tty(Some(prompt))?);
+        if password.0.is_empty() {
+            eprintln!("Passphrase must not be empty. Try again.");
+            continue;
+        }
+
+        let second_password = ZeroizeOnDrop(read_password_from_tty(Some(
+            "Enter same passphrase again: ",
```

Comment:
I'd call this variable `confirmation` instead of `second_password` which reads as if this is some new, unrelated password requested from the user.

Also, correct grammar would be

```suggestion
            "Enter the same passphrase again: ",
```

Reply:
Touché.